File: src-tauri/src/commands/staging.rs
// Shim layer for pollis_core::commands::staging. `stage_attachment` is
// hand-rolled to consume the IPC raw-byte body — the same shape
// `terminal_write` uses — because the whole point of the module is that these
// bytes never become a file. Edit pollis-core, not here.

#![allow(unused_imports)]
use std::sync::Arc;
use tauri::State;

use crate::error::{Error, Result};
use crate::state::AppState;
pub use pollis_core::commands::staging::*;

/// Take custody of an attachment's bytes and hand back the opaque handle the
/// renderer passes to `upload_media_staged`.
///
/// The payload is the raw IPC body — the frontend invokes with the
/// `Uint8Array` itself — so a pasted file crosses the boundary as bytes rather
/// than as the `Array.from(uint8) -> JSON number-array -> serde Vec<u8>`
/// round trip, which for a multi-megabyte screenshot is the difference between
/// a paste and a stall.
#[tauri::command]
pub async fn stage_attachment(request: tauri::ipc::Request<'_>) -> Result<StagedAttachment> {
    let bytes: Vec<u8> = match request.body() {
        tauri::ipc::InvokeBody::Raw(bytes) => bytes.clone(),
        tauri::ipc::InvokeBody::Json(_) => {
            return Err(Error::Other(anyhow::anyhow!(
                "stage_attachment: expected raw body, got json — the frontend must invoke \
                 with a Uint8Array"
            )));
        }
    };
    pollis_core::commands::staging::stage_attachment(bytes)
}

/// Release one staged attachment. Called when the user removes an attachment
/// card, and when a composer unmounts still holding staged bytes.
#[tauri::command]
pub fn discard_staged_attachment(staged_id: String) -> bool {
    pollis_core::commands::staging::discard_staged(&staged_id)
}

File: frontend/src/bridge/staging.ts
/**
 * Attachment staging bridge — `stage_attachment` and
 * `discard_staged_attachment` (`src-tauri/src/commands/staging.rs`).
 *
 * A pasted or dragged-in file arrives in the webview as a `File` with no
 * filesystem path, and the upload path needs somewhere to read it from. That
 * used to be a copy written into the OS temp directory under the file's own
 * name, which nothing ever deleted — so the plaintext of every file a user had
 * ever pasted was still there. These two calls replace it: the bytes go
 * straight into the backend's memory and come back out only as an upload.
 *
 * The whole point is that no path exists, so nothing here returns one.
 */

import { invoke } from "./invoke";

export interface StagedAttachment {
  /** Opaque handle; pass to `upload_media_staged` or `discardStagedAttachment`. */
  id: string;
  size_bytes: number;
}

/**
 * Hand the backend an attachment's bytes. Invoked with the `Uint8Array` as the
 * raw IPC body — not a JSON number array, which for a multi-megabyte
 * screenshot is the difference between a paste and a stall.
 */
export async function stageAttachment(
  bytes: Uint8Array,
): Promise<StagedAttachment> {
  return invoke<StagedAttachment>("stage_attachment", bytes);
}

/**
 * Release staged bytes the composer is no longer going to send — the user
 * removed the attachment card, or the composer unmounted still holding it.
 *
 * A successful upload releases its own entry, so this is not the only thing
 * standing between a paste and a leak; `lock`, `logout` and `wipe_local_data`
 * release everything, and process exit frees the lot with nothing on disk.
 */
export async function discardStagedAttachment(id: string): Promise<boolean> {
  return invoke<boolean>("discard_staged_attachment", { stagedId: id });
}
